### PR TITLE
implementation of btag sf

### PR DESCRIPTION
**Pull request by valeriadamante**
PR to implement the BTAg working point based (not shape!!) scale factors uncertainties

### CODE REVIEW

File: Common/bTagEffprod.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    dfw.Apply(Baseline.RecoHttCandidateSelection, config["GLOBAL"])
+    dfw.Apply(Baseline.RecoJetSelection)
+    df = dfw.df
+    pt_bins = Utilities.ListToVector([20,30,40,60,100,150,200,300,1000], "double")
+    eta_bins = Utilities.ListToVector([0,0.6,1.2,2.1,2.4], "double")
+    model = ROOT.RDF.TH2DModel("", "", pt_bins.size()-1, pt_bins.data(), eta_bins.size()-1, eta_bins.data())
+    hists = {}
+    for flav in [ 0, 4, 5]:
+        df = df.Define(f"Jet_eta_flavour{flav}", f"abs(v_ops::eta(Jet_p4[Jet_bCand && Jet_hadronFlavour=={flav}]))")
+        df = df.Define(f"Jet_pt_flavour{flav}", f"v_ops::pt(Jet_p4[Jet_bCand && Jet_hadronFlavour=={flav}])")
```
Comment (kandrosov) [inline Common/bTagEffprod.py:44]:
```suggestion
        df = df.Define(f"Jet_flavour{flav}_sel", f"Jet_bCand && Jet_hadronFlavour=={flav}")
        df = df.Define(f"Jet_eta_flavour{flav}", f"abs(v_ops::eta(Jet_p4[Jet_flavour{flav}_sel]))")
        df = df.Define(f"Jet_pt_flavour{flav}", f"v_ops::pt(Jet_p4[Jet_flavour{flav}_sel])")
```

File: Common/bTagEffprod.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        hist_name = f'jet_pt_eta_{flav}'
+        hists[hist_name] = df.Histo2D(model, f"Jet_pt_flavour{flav}", f"Jet_eta_flavour{flav}")
+        for wp, thr in wpValues.items():
+            df = df.Define(f"Jet_eta_WP{wp.name}_flavour{flav}", f"abs(v_ops::eta(Jet_p4[Jet_bCand && Jet_hadronFlavour=={flav} && Jet_btagDeepFlavB > {thr}]))")
+            df = df.Define(f"Jet_pt_WP{wp.name}_flavour{flav}", f"v_ops::pt(Jet_p4[Jet_bCand && Jet_hadronFlavour=={flav} && Jet_btagDeepFlavB > {thr}])")
```
Comment (kandrosov) [inline Common/bTagEffprod.py:49]:
```suggestion
            df = df.Define(f"Jet_WP{wp.name}_flavour{flav}_sel", f"Jet_flavour{flav}_sel && Jet_btagDeepFlavB > {thr}")
            df = df.Define(f"Jet_eta_WP{wp.name}_flavour{flav}", f"abs(v_ops::eta(Jet_p4[Jet_WP{wp.name}_flavour{flav}_sel]))")
            df = df.Define(f"Jet_pt_WP{wp.name}_flavour{flav}", f"v_ops::pt(Jet_p4[Jet_WP{wp.name}_flavour{flav}_sel])")
```

File: Common/bTagEffprod.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    isHH=False
+    with open(args.configFile, 'r') as f:
+        config = yaml.safe_load(f)
+    hists = bTagProdEff(args.inFile, config, args.sample, args.nEvents,args.evtIds)
+    fileToSave = ROOT.TFile(args.outFile, 'RECREATE')
+    for hist_name, hist in hists.items():
+        hist.SetTitle(hist_name)
+        hist.SetName(hist_name)
+        fileToSave.WriteTObject(hist.GetValue(), hist_name)
+    fileToSave.Close()
```
Comment (kandrosov) [inline Common/bTagEffprod.py:81]: run and print validation report checking for empty bins or bins with stat error for eff > thr. Where default thr e.g. 1%.

File: config/triggers_2018.yaml [hunk 4]
```diff
@@ -13,14 +13,12 @@ ditau:
...
-        cut: TrigObj_id == 15 && (TrigObj_filterBits&64)!=0
-      doMatching: True
-    - offline_obj:
-        type: Tau
-        cut: v_ops::pt(Tau_p4) > 40 && abs(v_ops::eta(Tau_p4)) < 2.1
-      online_obj:
-        cut: TrigObj_id == 15 && (TrigObj_filterBits&64)!=0 # for charged iso? and for HPS ??
+        cuts:
+          - cut: TrigObj_id == 15 && (TrigObj_filterBits&64)!=0
+          - cut: (TrigObj_filterBits&1)==0 && ( (TrigObj_filterBits&4)==0 || (TrigObj_filterBits&16)==0) && (TrigObj_pt>40 || ((TrigObj_filterBits&4)==0 && (TrigObj_filterBits&16)==0))
```
Comment (kandrosov) [inline config/triggers_2018.yaml:18]: fix to be aligned with the latest recommendations

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Common/bTagEffprod.py (hunk 1 above)
Comment (valeriadamante) [reply Common/bTagEffprod.py:44]: ok

File: Common/bTagEffprod.py (hunk 2 above)
Comment (valeriadamante) [reply Common/bTagEffprod.py:49]: ok

File: Common/bTagEffprod.py (hunk 3 above)
Comment (valeriadamante) [reply Common/bTagEffprod.py:81]: working on it

File: Common/HistValidation.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    for wp in wps:
+        key_num = f"{prefix}_{jet_fl}_{wp}"
+        hist_num = hist_file.Get(key_num)
+        for x_bin_number in range(1, hist_den.GetNbinsX()-1):
+            for y_bin_number in range(1, hist_den.GetNbinsY()-1):
+                bin_number = hist.GetBin(x_bin_number,y_bin_number)
+                num = hist_num.GetBinContent(bin_number)
+                den = hist_den.GetBinContent(bin_number)
+                c_low = ROOT.TEfficiency.ClopperPearson(den, num, 1-0.68, False)
+                c_up = ROOT.TEfficiency.ClopperPearson(den, num, 1-0.68, True)
```
Comment (kandrosov) [inline Common/HistValidation.py:45]:
```suggestion
                c_low = ROOT.TEfficiency.ClopperPearson(den, num, 0.68, False)
                c_up = ROOT.TEfficiency.ClopperPearson(den, num, 0.68, True)
```

File: Common/HistValidation.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                #print(f"low = {c_low}, up = {c_up}")
+                eff= num/den
+                rel_cl_size=(c_up- c_low)/(2*eff)
+                if round(rel_cl_size,2)>0.01:
```
Comment (kandrosov) [inline Common/HistValidation.py:49]:
```suggestion
                if rel_cl_size >= 0.01:
```

File: Common/HistValidation.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                  #""")
+    for x_bin_number in range(1, hist.GetNbinsX()-1):
+        for y_bin_number in range(1, hist.GetNbinsY()-1):
+            bin_number = hist.GetBin(x_bin_number,y_bin_number)
+            bin_content = hist.GetBinContent(bin_number)
+            if(bin_content==0):
+                print(f""" key = {key},
+                    bin_number = {bin_number}, bin_content = {bin_content},
+                    bin_center_X = {hist.GetXaxis().GetBinCenter(x_bin_number)},
+                    bin_center_y = {hist.GetYaxis().GetBinCenter(x_bin_number)}""")
```
Comment (kandrosov) [inline Common/HistValidation.py:29]: why this is needed?

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Common/HistValidation.py (hunk 5 above)
Comment (valeriadamante) [reply Common/HistValidation.py:45]: ok

File: Common/HistValidation.py (hunk 6 above)
Comment (valeriadamante) [reply Common/HistValidation.py:49]: ok

File: Common/HistValidation.py (hunk 7 above)
Comment (valeriadamante) [reply Common/HistValidation.py:29]: for personal check

Comment (valeriadamante) [reply Common/HistValidation.py:29]: I mean, we need to print when the bin content is 0!!

Comment (kandrosov) [reply Common/HistValidation.py:29]: but we have loop over all bins below. why to have `for key in keyList` and `for jet_fl in jet_flavors`?

File: Common/HistValidation.py (hunk 5 above)
Comment (kandrosov) [reply Common/HistValidation.py]: please, test that there are no warnings for the efficiencies with the correct CI definition.

File: Common/HistValidation.py (hunk 7 above)
Comment (valeriadamante) [reply Common/HistValidation.py]: Because we need to discriminate numerator from denumerator

File: Common/HistValidation.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+hist_file = ROOT.TFile("Corrections/data/BTV/2018_UL/btagEff.root","READ") # take as argument
+
+def GetKeyNames( self, dir = "" ):
+        self.cd(dir)
+        return [key.GetName() for key in ROOT.gDirectory.GetListOfKeys()]
+ROOT.TFile.GetKeyNames = GetKeyNames
+keyList = hist_file.GetKeyNames()
+
+nums = ["jet_pt_eta_0_Loose","jet_pt_eta_0_Medium","jet_pt_eta_0_Tight","jet_pt_eta_4_Loose","jet_pt_eta_4_Medium","jet_pt_eta_4_Tight","jet_pt_eta_5_Loose","jet_pt_eta_5_Medium","jet_pt_eta_5_Tight",]
+denums = ["jet_pt_eta_0", "jet_pt_eta_4", "jet_pt_eta_5"]
```
Comment (kandrosov) [inline Common/HistValidation.py:14]:
```suggestion

```

**Pull request merged**